packages/flutter_tools/lib/src/android/android_emulator.dart

```diff
@@ -50,18 +54,55 @@ class AndroidEmulator extends Emulator {
 
   @override
   Future<void> launch() async {
-    final Future<void> launchResult = processUtils.run(
-      <String>[getEmulatorPath(), '-avd', id],
-      throwOnError: true,
+    final Process process = await processUtils.start(
+      <String>[getEmulatorPath(androidSdk), '-avd', id],
     );
-    // The emulator continues running on a successful launch, so if it hasn't
-    // quit within 3 seconds we assume that's a success and just return. This
-    // means that on a slow machine, a failure that takes more than three
-    // seconds won't be recognized as such... :-/
-    return Future.any<void>(<Future<void>>[
-      launchResult,
-      Future<void>.delayed(const Duration(seconds: 3)),
+
+    // Record output from the emulator process.
+    final List<String> stdoutList = <String>[];
+    final List<String> stderrList = <String>[];
+    final StreamSubscription<String> stdoutSubscription = process.stdout
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .listen(stdoutList.add);
+    final StreamSubscription<String> stderrSubscription = process.stderr
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .listen(stderrList.add);
+    final Future<void> stdioFuture = waitGroup<void>(<Future<void>>[
+      stdoutSubscription.asFuture<void>(),
+      stderrSubscription.asFuture<void>(),
     ]);
+
+    // The emulator continues running on success, so we don't wait for the
+    // process to complete before continuing. However, if the process fails
+    // after the startup phase (3 seconds), then we only echo its output if
+    // its error code is non-zero and its stderr is non-empty.
+    bool earlyFailure = true;
+    unawaited(process.exitCode.then((int status) async {
+      if (status == 0) {
+        return;
+      }
+      // Make sure the process' stdout and stderr are drained.
+      await stdioFuture;
+      unawaited(stdoutSubscription.cancel());
+      unawaited(stderrSubscription.cancel());
+      if (!earlyFailure && stderrList.isEmpty) {
+        return;
```

Comment:
would it make sense to print message in this case as well?

Reply:
Yeah, I shuffled the prints around a bit, so that it'll be a bit more clear.